channels/rdpecam/client/camera.h

```diff
@@ -0,0 +1,191 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * MS-RDPECAM Implementation, main header file
+ *
+ * Copyright 2024 Oleg Turovski <oleg2104@hotmail.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FREERDP_CLIENT_CAMERA_H
+#define FREERDP_CLIENT_CAMERA_H
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <libswscale/swscale.h>
+#include <libavutil/imgutils.h>
+
+#include <winpr/wlog.h>
+
+#include <freerdp/api.h>
+#include <freerdp/types.h>
+
+#include <freerdp/client/channels.h>
+#include <freerdp/channels/log.h>
+#include <freerdp/channels/rdpecam.h>
+#include <freerdp/codecs.h>
+#include <freerdp/primitives.h>
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#define ECAM_PROTO_VERSION 0x02
+/* currently supporting 1 stream per device */
+#define ECAM_DEVICE_MAX_STREAMS 1
+#define ECAM_MAX_MEDIA_TYPE_DESCRIPTORS 256
+
+/* Allow to send up to that many unsolicited samples.
+ * For example, to support 30 fps with 250 ms round trip
+ * ECAM_MAX_SAMPLE_CREDITS has to be at least 8.
+ */
+#define ECAM_MAX_SAMPLE_CREDITS 8
+
+/* Having this hardcoded allows to preallocate and reuse buffer
+ * for sample responses. Excessive size is to make sure any sample
+ * will fit in, even with highest resolution.
+ */
+#define ECAM_SAMPLE_RESPONSE_BUFFER_SIZE (1024 * 4050)
+
+/* 4 Mbps max encoded bitrate seems to produce reasonably
+ * good quality with H264_RATECONTROL_VBR.
+ */
+#define ECAM_H264_ENCODED_BITRATE 4000000
+
+typedef struct s_ICamHal ICamHal;
+
+typedef struct
+{
+	IWTSPlugin iface;
+	IWTSListener* listener;
+	GENERIC_LISTENER_CALLBACK* hlistener;
+
+	/* HAL interface */
+	ICamHal* ihal;
+	char* subsystem;
+
+	BOOL initialized;
+	BOOL attached;
+
+	UINT32 version;
+	wHashTable* devices;
+
+} CameraPlugin;
+
+typedef struct
+{
+	CAM_MEDIA_FORMAT inputFormat;  /* camera side */
+	CAM_MEDIA_FORMAT outputFormat; /* network side */
+
+} CAM_MEDIA_FORMAT_INFO;
+
+typedef struct
+{
+	BOOL streaming;
+	CAM_MEDIA_FORMAT_INFO formats;
+	CAM_MEDIA_TYPE_DESCRIPTION currMediaType;
+
+	GENERIC_CHANNEL_CALLBACK* hSampleReqChannel;
+	INT nSampleCredits;
+	wStream* sampleRespBuffer;
+
+	H264_CONTEXT* h264;
+	struct SwsContext* sws;
+	int srcLineSizes[4];
+
+} CameraDeviceStream;
+
+#define streamInputFormat(stream) ((stream)->formats.inputFormat)
```

Comment:
better use `static INLINE` functions instead of `define` (typecheckes, ...)

Reply:
Fixed